eppo_core/src/events/event_dispatcher.rs

```diff
@@ -78,78 +76,96 @@ impl<T: EventQueue + Clone + Send + 'static> EventDispatcher<T> {
             }
+        }
+    }
+}
 
-            let event_delivery = event_delivery.clone();
-            let batch = event_queue.next_batch(QueuedEventStatus::Pending);
-            let queue = event_queue.clone();
-            // Send events
-            tokio::spawn({
-                async move {
-                    // Spawning a new task, so the main task can continue batching events and respond to
-                    // commands. At this point, batch_queue is guaranteed to have at least one event.
-                    let events = batch
-                        .iter()
-                        .map(|queued_event| queued_event.clone().event)
-                        .collect();
-                    let result = event_delivery.deliver(events).await;
-                    match result {
-                        Ok(response) => {
-                            let failed_event_uuids = response.failed_events;
-                            if !failed_event_uuids.is_empty() {
-                                warn!("Failed to deliver {} events", failed_event_uuids.len());
-                                let failed_events = batch
-                                    .into_iter()
-                                    .filter(|queued_event| {
-                                        failed_event_uuids.contains(&queued_event.event.uuid)
-                                    })
-                                    .collect();
-                                queue.mark_events_as_failed(failed_events);
-                            }
-                        }
-                        Err(err) => {
-                            warn!("Failed to deliver events: {}", err);
-                            queue.mark_events_as_failed(batch);
-                        }
+fn spawn_event_delivery<T: EventQueue + Clone + Send + 'static>(
+    batch: Vec<QueuedEvent>,
+    event_delivery: EventDelivery,
+    queue: T,
+) {
+    if batch.is_empty() {
+        // nothing to deliver
+        return;
+    }
+    // Send events
+    tokio::spawn({
+        async move {
+            // Spawning a new task, so the main task can continue batching events and respond to
+            // commands. At this point, batch is guaranteed to have at least one event.
+            let events = batch
+                .iter()
+                .map(|queued_event| queued_event.clone().event)
+                .collect();
+            let result = event_delivery.deliver(events).await;
+            match result {
+                Ok(response) => {
+                    let failed_event_uuids = response.failed_events;
+                    if !failed_event_uuids.is_empty() {
+                        warn!("Failed to deliver {} events", failed_event_uuids.len());
+                        let failed_events = batch
+                            .into_iter()
+                            .filter(|queued_event| {
+                                failed_event_uuids.contains(&queued_event.event.uuid)
```

Comment:
`failed_event_uuids` is a vector, so `contains()` is O(n). Because this operation is wrapped in another O(n) (`filter`), that totals to O(n^2). It's cheaper to convert `failed_event_uuids` to a hashset and then do the check (I think we can simply change the type in `EventDeliveryResponse`, so it deserializes directly into set)

Reply:
good idea!